core/src/org/pentaho/di/core/extension/ExtensionPointMap.java

```diff
@@ -105,38 +103,91 @@ public void addExtensionPoint( PluginInterface extensionPointPlugin ) {
    * @param extensionPointPlugin
    */
   public void removeExtensionPoint( PluginInterface extensionPointPlugin ) {
-    for ( String id : extensionPointPlugin.getIds() ) {
-      extensionPointPluginMap.remove( extensionPointPlugin.getName(), id );
+    lock.writeLock().lock();
+    try {
+      for ( String id : extensionPointPlugin.getIds() ) {
+        extensionPointPluginMap.remove( extensionPointPlugin.getName(), id );
+      }
+    } finally {
+      lock.writeLock().unlock();
     }
   }
 
   /**
-   * Retrieves the
-   * 
-   * @param id
-   * @return
-   */
-  public Map<String, ExtensionPointInterface> get( String id ) {
-    return Maps.transformValues( extensionPointPluginMap.row( id ),
-      Suppliers.<ExtensionPointInterface>supplierFunction() );
-  }
-
-  /**
    * Reinitialize the extension point plugins map
    */
   public void reInitialize() {
-    extensionPointPluginMap = HashBasedTable.create();
-    final PluginRegistry registry = PluginRegistry.getInstance();
-    List<PluginInterface> extensionPointPlugins = registry.getPlugins( ExtensionPointPluginType.class );
-    for ( PluginInterface extensionPointPlugin : extensionPointPlugins ) {
-      addExtensionPoint( extensionPointPlugin );
+    lock.writeLock().lock();
+    try {
+      extensionPointPluginMap = HashBasedTable.create();
+      final PluginRegistry registry = PluginRegistry.getInstance();
+      List<PluginInterface> extensionPointPlugins = registry.getPlugins( ExtensionPointPluginType.class );
+      for ( PluginInterface extensionPointPlugin : extensionPointPlugins ) {
+        addExtensionPoint( extensionPointPlugin );
+      }
+    } finally {
+      lock.writeLock().unlock();
     }
   }
 
   Supplier<ExtensionPointInterface> createLazyLoader( PluginInterface extensionPointPlugin ) {
     return Suppliers.memoize( new ExtensionPointLoader( extensionPointPlugin ) );
   }
 
+  /**
+   * Call the extension point(s) corresponding to the given id
+   *
+   * This iteration was isolated here to protect against ConcurrentModificationException using PluginRegistry's lock
+   *
+   * @param log     log channel to pass to extension point call
+   * @param id      the id of the extension point interface
+   * @param object  object to pass to extension point call
+   */
+  public void callExtensionPoint( LogChannelInterface log, String id, Object object ) throws KettleException {
+    lock.writeLock().lock();
```

Comment:
Since you're not changing the map here, shouldn't this get the read lock instead of the write lock?